### PR TITLE
#1153 - mkdir supports creation of parent directories

### PR DESCRIPTION
**Pull request by alachata-qore**

### CODE REVIEW

File: lib/ql_file.qpp [hunk 1]
```diff
@@ -667,6 +667,7 @@ nothing umask() [flags=RUNTIME_NOOP;dom=PROCESS] {
 //! Creates a directory, optionally specifying the mode
 /** @param path the path to the directory to create
     @param mode the file mode of the new directory (which will be AND'ed with the umask)
+    @param parents if True also parrent directories are created
```
Comment (davidnich) [inline lib/ql_file.qpp]: please fix the typo: `parrent` -> `parent`

File: lib/ql_file.qpp [hunk 2]
```diff
@@ -678,12 +679,60 @@ if (mkdir("/tmp/newdir", 0755))
...
+int mkdir(string path, softint mode = 0777, bool parents = False) [dom=FILESYSTEM] {
+    const size_t len = strlen(path->getBuffer());
+    char _path[PATH_MAX];
+    char *p;
+
+    errno = 0;
+
 #ifdef _Q_WINDOWS
-   return _mkdir(path->getBuffer());
+    char sep = '\\';
```
Comment (davidnich) [inline lib/ql_file.qpp]: this is already defined in common.h, please do not redefine - use `QORE_DIR_SEP`

File: lib/ql_file.qpp [hunk 3]
```diff
@@ -678,12 +679,60 @@ if (mkdir("/tmp/newdir", 0755))
...
 #else
-   return mkdir(path->getBuffer(), mode);
+    char sep = '/';
 #endif
+    /* Copy string so its mutable */
```
Comment (davidnich) [inline lib/ql_file.qpp]: please use c++ line comments `//`

**Review by davidnich — CHANGES_REQUESTED** [review]
if `parents` = True then you should use `QoreDir::create()` instead of your own implementation

File: lib/ql_file.qpp [hunk 4]
```diff
@@ -667,8 +669,9 @@ nothing umask() [flags=RUNTIME_NOOP;dom=PROCESS] {
 //! Creates a directory, optionally specifying the mode
 /** @param path the path to the directory to create
     @param mode the file mode of the new directory (which will be AND'ed with the umask)
+    @parents if True then parent directories are created
 
-    @return 0 if no error occurred; -1 if an error occurred, in which case errno() and/or strerror() can be used to retrieve the error
+    @return 0 or count of newly created directories if no error occured; -1 if an error occurred, in which case errno() and/or strerror() can be used to retrieve the error
```
Comment (davidnich) [inline lib/ql_file.qpp]: count -> "the count" and occured -> "occurred"

**Review by davidnich — COMMENTED** [review]
please also add a release note `doxygen/lang/900_release_notes.dox.tmpl` - after "new function" add "updated functions" and list the changes to this function

also please add a `@since %Qore 0.8.13 added the <i>parents</i> parameter`  line to the `mkdir()` function documentation

also the return value documentation for `mkdir()` should reflect that fact that the count of newly created directories is only returned when <i>parents</i> = @ref Qore::True "True", otherwise the return value for a successful call remains `0`

**Comment by alachata-qore** [comment]
Please check now.

File: doxygen/lang/900_release_notes.dox.tmpl [hunk 5]
```diff
@@ -70,6 +70,8 @@
       - <number>::format(int, string, string)
     - new functions:
       - @ref Qore::parse_int()
+    - updated functions:
+      - @ref Qore::mkdir(string path, softint mode = 0777, bool paretns = False)
```
Comment (davidnich) [inline doxygen/lang/900_release_notes.dox.tmpl]: please fix the typo `paretns`

File: lib/ql_file.qpp [hunk 6]
```diff
@@ -667,8 +669,9 @@ nothing umask() [flags=RUNTIME_NOOP;dom=PROCESS] {
 //! Creates a directory, optionally specifying the mode
 /** @param path the path to the directory to create
     @param mode the file mode of the new directory (which will be AND'ed with the umask)
+    @parents if True then parent directories are created
 
-    @return 0 if no error occurred; -1 if an error occurred, in which case errno() and/or strerror() can be used to retrieve the error
+    @return if no error occured returns 0 or the count of newly created directory when parents = @ref Qore::True "True"; -1 if an error occurred, in which case errno() and/or strerror() can be used to retrieve the error
```
Comment (davidnich) [inline lib/ql_file.qpp]: typo - occured -> `occurred`

**Review by davidnich — CHANGES_REQUESTED** [review]

**Pull request merged**